Implement keyboard API

koxu1996 opened the pull request:
### Base API

Functions:
 - `p5.KeyIsDown(code string) bool` - checks if given key is already pressed

Variables:
- `p5.KeyIsPressed bool` - whether any key is pressed now
- `p5.Key string` - name of last pressed/typed key 

### Callbacks

Callbacks can be set with following functions:
- `p5.SetKeyPressedCallback(f KeyEventFunc)`
- `p5.SetKeyTypedCallback(f KeyEventFunc)`
- `p5.SetKeyReleasedCallback(f KeyEventFunc)`

---

**Caveats:**

- There is no way to detect lower case characters, so we always use upper case. Bear that in mind, especially when working with _KeyTyped_ event.

---

@sbinet I am not sure if the code is safe, so please take a look. I used your examples _key-snake_ and _key-pressed_ from https://github.com/go-p5/p5/pull/52. Feel free to suggest changes, this is just proposal :grinning:

sbinet commented:
thanks for the PR.

I like you `map`-based approach. (I was initially thinking of using a `uint32` mask, but, well, if the `map` approach is efficient enough, who cares)

I think I'd rather have only one `struct` that holds all the callbacks than increasing the number of globals.
_e.g.:_
```go
package p5

type Keyboard struct {
    down    map[string]struct{}
    release map[key.Event]struct{}
    last    string

    Pressed  KeyCallback
    Typed    KeyCallback
    Released KeyCallback
}

func WithKeyCallback(cbk Keyboard) Option {
    return func(p *Proc) { p.Key = cbk }
}
```

and:
```go
package main

func main() {
    p5.Run(setup, draw, p5.WithKeyCallback(p5.Keyboard{
        Pressed: keyPressed,
        ...
    })
}
```

what do you think?

koxu1996 commented:
1. I made little refactoring, now callbacks and global variables are moved to single struct.

2. Regarding your callback approach, user have to pass all 3 callbacks otherwise `SIGSEGV` is thrown. I would rather allow setting each individually. Additionally your example is using `Proc.Key` which is missing, and I am confused if you want to store callbacks in `Keyboard` or `Proc` structure.

3. I don't see how `uint32` mask could be used; for storing keys map is suitable and efficient container. Regarding following code:

> down    map[string]struct{}
> release map[key.Event]struct{}

- map with bool is recommended for set-like data - see https://blog.golang.org/maps#TOC_4
- we cannot use `key.Event` as key, because we could not search this map in simple way

sbinet commented:
> Regarding your callback approach, user have to pass all 3 callbacks otherwise SIGSEGV is thrown. 

not necessarily.
this would need some plumbing in `p5.Run(...)`, though:
```go
func Run(setup, draw Func, opts ...Option) {
...
    for _, opt := range opts {
        opt(gproc)
    }
    if gproc.Key.Pressed == nil {
        gproc.Key.Pressed = func() {}
    }
    ...
}
```

> Additionally your example is using Proc.Key which is missing

I was indeed using notation from my previous PR :)
I was also probably beeing overeager and starting to reduce the number of globals being used in the `p5` package.

> map with bool is recommended for set-like data - see https://blog.golang.org/maps#TOC_4

this blog entry is a bit old :)
I think nowadays the `map[T]struct{}` is prefered as it brings the same semantics than a `map[T]bool` without the extra memory foot print of the useless boolean associated value (the empty struct `struct{}` will be optimized away).

> we cannot use key.Event as key, because we could not search this map in simple way

you're right.

(I am sorry I'll have to disappear for a week w/o much internet connectivity and leave this PR unreviewed for the time being. I'll review it further coming back from holidays)

koxu1996 commented:
Okay, I replaced `map[string]bool` with `map[string]struct{}`. I moved callbacks into `Proc` as you wished and now they are passed via single `Option`:

```
func main() {
	p5.Run(setup, draw, p5.WithKeyCallback(p5.KeyCb{
		Pressed:  keyPressed,
		Typed:    keyTyped,
	}))
	...
```
Is it fine?

koxu1996 commented:
@sbinet Do you have time to look at it?

sbinet commented on proc.go:
```diff
@@ -79,9 +79,12 @@ var _ gioWindow = (*app.Window)(nil)
+	Setup       Func
+	Draw        Func
+	Mouse       Func
+	KeyPressed  KeyEventFunc
+	KeyTyped    KeyEventFunc
+	KeyReleased KeyEventFunc
```
instead of "exploding" the `KeyCb` fields into 3 different fields in `Proc`, perhaps just add a `Key KeyCb` field to `Proc` ?

sbinet commented on keyboard.go:
```diff
@@ -0,0 +1,33 @@
+// KeyEventFunc is the type of key functions users provide to p5.
+type KeyEventFunc func(key.Event)
+
+type KeyCb struct {
```
this new type probably needs a bit of documentation.
also, as we spelled out the func-option `WithKeyCallback`, what about renaming this type `KeyCallback` ?
(alternatively, we could go with `WithKeyCbk / KeyCbk`)

the meaning of (and their semantical nuance) the `Pressed` and `Typed` probably need to be documented and/or explained.
this can be done as part of the documentation of each of these fields.

sbinet commented on api.go:
```diff
@@ -145,3 +145,8 @@ func ReadImage(fname string) (image.Image, error) {
 	gproc.DrawImage(img, x, y)
 }
+
+// KeyIsDown checks if given key is already pressed.
```
```suggestion
// KeyIsDown checks whether a given key is already pressed.
```

sbinet commented on api.go:
```diff
@@ -145,3 +145,8 @@ func ReadImage(fname string) (image.Image, error) {
 }
+
+// KeyIsDown checks if given key is already pressed.
+func KeyIsDown(code string) bool {
```
```suggestion
func KeyIsDown(name string) bool {
```

sbinet commented on keyboard.go:
```diff
@@ -0,0 +1,33 @@
+	lastKeyTyped    string
+}
+
+// KeyIsDown checks if given key is already pressed.
+func (p *Proc) KeyIsDown(code string) bool {
```
```suggestion
// KeyIsDown checks whether a given key is already pressed.
func (p *Proc) KeyIsDown(name string) bool {
```

sbinet commented on keyboard.go:
```diff
@@ -0,0 +1,33 @@
+
+// KeyIsDown checks if given key is already pressed.
+func (p *Proc) KeyIsDown(code string) bool {
+	if _, ok := Keyboard.downKeys[code]; ok {
+		return true
+	}
+	return false
```
```suggestion
	_, ok := Keyboard.downKeys[code]
	return ok
```

sbinet reviewed (commented):
hi,

apologies for the belated answer.
please find my comments inlined.

do you have an idea on a testing strategy for this new feature?
(something testable through the usual `go test` scaffolding)

also: could you rebase off the latest `main`?

thanks.

sbinet commented:
ping?

koxu1996 commented:
Sorry, I am quite busy nowadays and I am no longer interested in using _go-p5_.
